src/Open3D/GUI/SceneWidget.cpp

```diff
@@ -124,11 +144,321 @@ class CameraControls {
+            case DragType::TWO_FINGER:
+                // Zoom out is "push away" or up, is a positive value for
+                // two-finger scrolling, so we need to invert dy.
+                dist = float(-dy) * 0.005f * modelSize_;
+                break;
+            case DragType::WHEEL:  // actual mouse wheel, same as two-fingers
+                dist = float(-dy) * 0.1f * modelSize_;
+                break;
+        }
+
+        if (dragType == DragType::MOUSE) {
+            Dolly(dist, matrixAtMouseDown_);  // copies the matrix
+        } else {
+            Dolly(dist, matrix_);
+        }
+    }
+
+    // Note: we pass `matrix` by value because we want to copy it,
+    //       as translate() will be modifying it.
+    virtual void Dolly(float zDist, visualization::Camera::Transform matrix) {
+        // Dolly is just moving the camera forward. Filament uses right as +x,
+        // up as +y, and forward as -z (standard OpenGL coordinates). So to
+        // move forward all we need to do is translate the camera matrix by
+        // dist * (0, 0, -1). Note that translating by camera_->GetForwardVector
+        // would be incorrect, since GetForwardVector() returns the forward
+        // vector in world space, but the translation happens in camera space.)
+        // Since we want trackpad down (negative) to go forward ("pulling" the
+        // model toward the viewer) we need to negate dy.
+        auto forward = Eigen::Vector3f(0, 0, -zDist);  // zDist * (0, 0, -1)
+        matrix.translate(forward);
+        matrix_ = matrix;
+    }
+
+private:
+    visualization::Camera::Transform matrix_;
+
+    visualization::Camera::Transform matrixAtMouseDown_;
+    Eigen::Vector3f centerOfRotationAtMouseDown_;
+
+protected:
+    Size viewSize_;
+    double modelSize_ = 100.0;
+    geometry::AxisAlignedBoundingBox modelBounds_;
+    Eigen::Vector3f centerOfRotation_;
+};
+
+class LightDirControl : public MatrixControl {
+    using Super = MatrixControl;
+public:
+    LightDirControl(visualization::Camera *camera)
+    : camera_(camera) {
+    }
+
+    void SetDirectionalLight(visualization::Scene *scene,
+                             visualization::LightHandle dirLight) {
+        scene_ = scene;
+        dirLight_ = dirLight;
+    }
+
+    void Rotate(int dx, int dy) override {
+        Eigen::Vector3f up = camera_->GetUpVector();
+        Eigen::Vector3f right = -camera_->GetLeftVector();
+        RotateWorld(-dx, -dy, up, right);
+        UpdateMouseDragUI();
+    }
+
+    void StartMouseDrag() {
+        lightDirAtMouseDown_ = scene_->GetLightDirection(dirLight_);
+        auto identity = visualization::Camera::Transform::Identity();
+        Super::SetMouseDownInfo(identity, {0.0f, 0.0f, 0.0f});
+
+        for (auto &o : uiObjs_) {
+            scene_->RemoveGeometry(o.handle);
+        }
+
+        auto dir = scene_->GetLightDirection(dirLight_);
+
+        // TODO: it seems that SetEntityTransform moves an object to be
```

Comment:
It seems to be. If I comment out the transforms lines, and move adjust the vertices, everything gets moved back to the origin as soon as I add transform, even the identity transform, but it renders where I expect if I don't add a transform (and comment out that part in AddGeometry(). The moving part is definely Filament, because FilamentScene::AddGeometry() isn't doing, and the vertices aren't centered around the origin.

Reply:
Hm, I think I'll add task to Asana for myself, to check problems with transforms